coderd/workspaceusage/tracker.go

```diff
@@ -0,0 +1,229 @@
+func New(s Store, opts ...Option) *Tracker {
+	hb := &Tracker{
+		log:      slog.Make(sloghuman.Sink(os.Stderr)),
+		m:        &uuidSet{},
+		s:        s,
+		tickCh:   nil,
+		stopTick: nil,
+		stopCh:   make(chan struct{}),
+		doneCh:   make(chan struct{}),
+		flushCh:  nil,
+	}
+	for _, opt := range opts {
+		opt(hb)
+	}
+	if hb.tickCh == nil && hb.stopTick == nil {
+		ticker := time.NewTicker(DefaultFlushInterval)
+		hb.tickCh = ticker.C
+		hb.stopTick = ticker.Stop
+	}
+	return hb
+}
+
+type Option func(*Tracker)
+
+// WithLogger sets the logger to be used by Tracker.
+func WithLogger(log slog.Logger) Option {
+	return func(h *Tracker) {
+		h.log = log
+	}
+}
+
+// WithFlushInterval allows configuring the flush interval of Tracker.
+func WithFlushInterval(d time.Duration) Option {
+	return func(h *Tracker) {
+		ticker := time.NewTicker(d)
+		h.tickCh = ticker.C
+		h.stopTick = ticker.Stop
+	}
+}
+
+// WithFlushChannel allows passing a channel that receives
+// the number of marked workspaces every time Tracker flushes.
+// For testing only and will panic if used outside of tests.
+func WithFlushChannel(c chan int) Option {
+	if flag.Lookup("test.v") == nil {
+		panic("developer error: WithFlushChannel is not to be used outside of tests.")
+	}
+	return func(h *Tracker) {
+		h.flushCh = c
+	}
+}
+
+// WithTickChannel allows passing a channel to replace a ticker.
+// For testing only and will panic if used outside of tests.
+func WithTickChannel(c chan time.Time) Option {
+	if flag.Lookup("test.v") == nil {
+		panic("developer error: WithTickChannel is not to be used outside of tests.")
+	}
+	return func(h *Tracker) {
+		h.tickCh = c
+		h.stopTick = func() {}
+	}
+}
+
+// Add marks the workspace with the given ID as having been used recently.
+// Tracker will periodically flush this to its configured Store.
+func (wut *Tracker) Add(workspaceID uuid.UUID) {
+	wut.m.Add(workspaceID)
+}
+
+// flush updates last_used_at of all current workspace IDs.
+// If this is held while a previous flush is in progress, it will
+// deadlock until the previous flush has completed.
+func (wut *Tracker) flush(now time.Time) {
+	// Copy our current set of IDs
+	ids := wut.m.UniqueAndClear()
+	count := len(ids)
+	if wut.flushCh != nil { // only used for testing
+		defer func() {
+			wut.flushCh <- count
+		}()
+	}
+	if count == 0 {
+		wut.log.Debug(context.Background(), "nothing to flush")
+		return
+	}
+
+	// Set a short-ish timeout for this. We don't want to hang forever.
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	// nolint: gocritic // system function
+	authCtx := dbauthz.AsSystemRestricted(ctx)
+	wut.flushLock.Lock()
+	defer wut.flushLock.Unlock()
+	if err := wut.s.BatchUpdateWorkspaceLastUsedAt(authCtx, database.BatchUpdateWorkspaceLastUsedAtParams{
+		LastUsedAt: now,
+		IDs:        ids,
+	}); err != nil {
+		wut.log.Error(ctx, "failed updating workspaces last_used_at", slog.F("count", count), slog.Error(err))
```

Comment:
Is there any retry mechanism if the batch fails?

Reply:
There is a built-in assumption here that any failure of a single batch can be ignored as it gets retried frequently. I suppose consecutive failures should surface some form of error somewhere. I wonder if the database health page could show a summary of queries that have failed multiple consecutive executions?

Reply:
> There is a built-in assumption here that any failure of a single batch can be ignored as it gets retried frequently. I suppose consecutive failures should surface some form of error somewhere

I might have missed it, but it would be cool to leave a comment stating this.

Reply:
Done